### ISSUE TITLE
SVG (sprite) in Frame not rendered in Firefox

### ISSUE THREAD

**Issue by stgm** [hotwired/turbo#123]
**Context**

Firefox has this problem, but Safari and Chrome don't.

**Problem**

I have some SVG sprite icons that are not displayed when they are part of a dynamically loaded Turbo Frame. The embedded SVG looks like this:

```html
<svg class="bi" width="20" height="20" fill="currentColor">
    <use xlink:href="/icons/bootstrap-icons.svg#envelope"/>
</svg>
```

The sprite has already been loaded before activating the new content, because the page contains a couple of icons that are  displayed correctly.

**Solution**

The following change in Turbo will solve my problem, but I don't know what the consequences are. In effect, `createContextualFragment` is replaced by a `DOMParser` call as inspired by the Turbo Drive loading code.

```js
function fragmentFromHTML(html) {
  if (html) {
    const { documentElement } = new DOMParser().parseFromString(html, "text/html");
    return documentElement;
    // const foreignDocument = document.implementation.createHTMLDocument();
    // return foreignDocument.createRange().createContextualFragment(html);
  }
}
```

Please let me know if I can be of further assistance!

**Comment by guitarneck** [comment]
Hi,
Try with adding this attributes to your svg tag `version="1.1" xmlns="http://www.w3.org/2000/svg"`

**Comment by stgm** [comment]
Thanks @guitarneck! Unfortunately that doesn't change Firefox's behaviour. Here's what the markup looks like, so you might verify that I did it correctly:

```html
<svg class="bi" width="20" height="20" fill="currentColor" version="1.1" xmlns="http://www.w3.org/2000/svg">
    <use xlink:href="/icons/bootstrap-icons.svg#envelope"/>
</svg>
```

**Comment by stgm** [comment]
For what it's worth, previously, without Turbo, the same content was loaded into the DOM using `document.getElementById('modal-browser-header').innerHTML =`, which also works fine in Firefox.

**Comment by tleish** [comment]
@stgm —  can you create a working/no-working demo of the issue somewhere?

**Comment by stgm** [comment]
Sure @tleish, here's a page that automatically lazy-loads a Turbo frame containing an icon: 

https://stgm.github.io/firefox-sprite-bug/

There's also a single heart icon on the original page just to show that the SVG is working in general.

**Comment by tleish** [comment]
It may have something to do with with how turbo encapsulates the web component and how Firefox deals with a closed [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM)?  Just a guess.

As a workaround you could try using the full qualified url to the SVG file.

```html
<turbo-frame id="more-icons">
  <svg class="bi" width="32" height="32" fill="currentColor">
    <use href="https://stgm.github.io/firefox-sprite-bug/bootstrap-icons.svg#shop"/>
  </svg>
</turbo-frame>
```

BTW, SVG 2 [deprecated the xlink namespace](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/xlink:href), so instead of `xlink:href` you should use just [href](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/href).

```diff
-   <use xlink:href="https://stgm.github.io/firefox-sprite-bug/bootstrap-icons.svg#shop"/>
+   <use href="https://stgm.github.io/firefox-sprite-bug/bootstrap-icons.svg#shop"/>
```

**Comment by stgm** [comment]
Oh wow, specifying the absolute URL does help! Thanks for pointing that out. It's not something I _really_ want to use as a workaround though ;-)

It seems that parsing the fragment from the current document, instead of from the "foreign" document, also works.

```js
document.createRange().createContextualFragment(html);
```

**Comment by guitarneck** [comment]
Interesting...

**document.implementation** returns a DOMImplementation object.
[https://developer.mozilla.org/en-US/docs/Web/API/DOMImplementation](https://developer.mozilla.org/en-US/docs/Web/API/DOMImplementation)

Where there's two methods of our interest :
_DOMImplementation.createDocument() : XMLDocument_
and
_DOMImplementation.createHTMLDocument() : Document_

But it is says :
_The Document interface describes the common properties and methods for any kind of document. Depending on the document's type (e.g. HTML, XML, SVG, …), a larger API is available: **HTML documents, served with the "text/html" content type, also implement the HTMLDocument interface, whereas XML and SVG documents implement the XMLDocument interface**._

[https://developer.mozilla.org/en-US/docs/Web/API/Document](https://developer.mozilla.org/en-US/docs/Web/API/Document)

So **createDocument()** should do the trick for SVG, not **createHTMLDocument()**.

**Comment by guitarneck** [comment]
Mistake ans evidence !!!
**document.implementation.createHTMLDocument()** create a document with "about:blank" url address.
so `xlink:href="/icons/bootstrap-icons.svg#envelope"` won't solve, because of relative address ! But a static link will.

**Comment by stgm** [comment]
Well, yes and no. The document is created with a `baseURI` of `about:blank`, but as soon as it is added to the parent document its `baseURI` is updated accordingly.

Importing into the document is done using `document.importNode` in the function `activateElement` in `frame_controller.ts`.

Inspecting the DOM properties doesn't reveal any differences between icons that are visible (in the main doc) or invisible (loaded via Turbo frames).

**Comment by stgm** [comment]
@guitarneck Do you have the same problem for other things than SVG elements? I ask because in your PR you say "breaks any relative links in the loaded fragment". It seems that the `img` tag works fine, for example.

**Comment by guitarneck** [comment]
@stgm with `<img..>` it works well (png). `<embed..>` with svg needs some changes, as it not not a valid _image/svg+xml_, like so :
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg xmlns="http://www.w3.org/2000/svg"
  xmlns:svg="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  version="1.1"
  class="bi" width="32" height="32" fill="currentColor">
  <use xlink:href="bootstrap-icons.svg#shop"/>
</svg>
```
Adding `<image />` inside the svg works with png :
[https://developer.mozilla.org/fr/docs/Web/SVG/Element/image](https://developer.mozilla.org/fr/docs/Web/SVG/Element/image)

`importNode` made a clone of `foreignDocument`.
As I said, `createDocument `should be used for svg, `createHTMLDocument` (I'll test it). But adding `<base../>` for Firefox do the stuff. Then `foreignDocument ` gets the right baseURI.

**Comment by stgm** [comment]
Closed by #133, which doesn't use `createContextualFragment()` anymore. Sprites now show in frame-loaded content.
